src/test/java/io/cdap/plugin/batch/ETLFTPTestRun.java

```diff
@@ -0,0 +1,167 @@
+  }
+
+  @After
+  public void stop() {
+    if (ftpServer != null) {
+      ftpServer.stop();
+    }
+  }
+
+  @Test
+  public void testFTPBatchSource() throws Exception {
+    ETLStage source = new ETLStage("source", new ETLPlugin(
+      "FTP",
+      BatchSource.PLUGIN_TYPE,
+      ImmutableMap.<String, String>builder()
+        .put(PATH, String.format("ftp://%s:%s@localhost:%d%s",
+                                 USER, PWD, port, folder.getAbsolutePath()))
+        .put(IGNORE_NON_EXISTING_FOLDERS, "false")
+        .put("referenceName", "ftp")
+        .build(),
+      null));
+
+    String outputDatasetName = "testing-ftp-source";
+    ETLStage sink = new ETLStage("sink", MockSink.getPlugin(outputDatasetName));
+
+    ETLBatchConfig etlConfig = ETLBatchConfig.builder()
+      .addStage(source)
+      .addStage(sink)
+      .addConnection(source.getName(), sink.getName())
+      .build();
+
+    AppRequest<ETLBatchConfig> appRequest = new AppRequest<>(DATAPIPELINE_ARTIFACT, etlConfig);
+    ApplicationId appId = NamespaceId.DEFAULT.app("FTPBatchSource");
+
+    ApplicationManager appManager = deployApplication(appId, appRequest);
+
+    WorkflowManager workflowManager = appManager.getWorkflowManager(SmartWorkflow.NAME);
+    workflowManager.startAndWaitForRun(ProgramRunStatus.COMPLETED, 30, TimeUnit.SECONDS);
+
+    DataSetManager<Table> outputManager = getDataset(outputDatasetName);
+    List<StructuredRecord> output = MockSink.readOutput(outputManager);
+
+    Assert.assertEquals("Expected records", 1, output.size());
+    Set<String> outputValue = new HashSet<>();
+    for (StructuredRecord record : output) {
+      outputValue.add(record.get("body"));
+    }
+    Assert.assertTrue(outputValue.contains("Hello World"));
+  }
+
+  private void testHelper(Map<String, String> properties, Map<String, String> runTimeProperties) throws Exception {
+    ETLStage source = new ETLStage("source", new ETLPlugin(
+      "FTP", BatchSource.PLUGIN_TYPE, properties, null));
+
+    String outputDatasetName = "testing-ftp-source-with-macro";
+    ETLStage sink = new ETLStage("sink", MockSink.getPlugin(outputDatasetName));
+
+    ETLBatchConfig etlConfig = ETLBatchConfig.builder()
+      .addStage(source)
+      .addStage(sink)
+      .addConnection(source.getName(), sink.getName())
+      .build();
+
+    ApplicationManager appManager = deployETL(etlConfig, "FTPWithMacro");
+    runETLOnce(appManager, runTimeProperties);
+
+    DataSetManager<Table> outputManager = getDataset(outputDatasetName);
+    List<StructuredRecord> output = MockSink.readOutput(outputManager);
+
+    Assert.assertEquals("Expected records", 1, output.size());
+    Set<String> outputValue = new HashSet<>();
+    for (StructuredRecord record : output) {
+      outputValue.add(record.get("body"));
+    }
+    Assert.assertTrue(outputValue.contains("Hello World"));
+  }
+
+  @Test
+  public void testFTPBatchSourceWithMacro() throws Exception {
```

Comment:
please add negative tests with invalid path

Reply:
> please add negative tests with invalid path

We already added tests with invalid path. 

https://github.com/AdaptiveScale/ftp-plugins/blob/c40deec6eea51681e1c6ab1c4812dd0f29420e32/src/test/java/io/cdap/plugin/batch/source/FTPBatchSourceTest.java#L46-L59

Reply:
`testFTPPathWithSpecialCharactersInAuth()` is testing password with special characters. I think @CuriousVini is talking a negative test (i.e. test is expected to fail) where the FTP path is invalid.

Reply:
Ah okay, I understood. In the last update I added a test with invalid path.